DOC: Document ma.filled behavior with non-scalar fill_value

kritisingh1 opened the pull request:
<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->
Closes #4363
xref #13241 #10504
<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->
@mattip for now, I have just documented the behavior with non-scalar fill values. The PRs xref'ed above talked about deprecating the non-scalar nature(optional) of fill value. Do we really need to deprecate if we are already mentioning that, that behavior is not supported and throwing a meaningful error?
Also, should I include examples of what fill_values with input array combination could be risky?

kritisingh1 committed:
DOC: Document ma.filled behavior with non-scalar fill_value

mattip commented:
I think examples would be nice. We should also standardize the `Defalt is ...` stanza, altough the [nupydoc guide](https://numpydoc.readthedocs.io/en/latest/format.html#sections) doesn't give clear guidance to what is the One Best Way. Maybe it appears somewhere else?

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -602,8 +602,10 @@ def filled(a, fill_value=None):
         An input object.
-    fill_value : scalar, optional
-        Filling value. Default is None.
+    fill_value : Filling value, optional. Can be
```
First line should be an approximation of the type, not a description. `array_like, optional` would probably be correct.

eric-wieser reviewed (dismissed):
I'm not sure we want to document this, as remarked in the other issues you xref

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -6206,7 +6210,9 @@ def filled(self, fill_value=None):
         fill_value : scalar, optional
             The value to use for invalid entries (None by default).
-            If None, the `fill_value` attribute is used instead.
+            If None, the `fill_value` attribute is used instead. Can
```
I think this sentence can go last.

mattip commented:
We should implement the deprecation warning as in #13421, and add the examples from there.

kritisingh1 committed:
Adds examples

kritisingh1 committed:
Fix tests

kritisingh1 commented:
The test I have commented out is the one which fails due to `Attribute error: type object tuple has no attribute shape`. I am not sure what should be done for this case. In my opinion, it would be too naive to add a condition in the `if statement` for passing one test. Instead we can modify the test in a way that it does not contain the commented out line and instead write deprecation tests for checking the behavior when fill_value is non-scalar non-broadcastable. Let me know what would be better.

seberg commented on numpy/ma/core.py:
```diff
@@ -3708,7 +3730,13 @@ def filled(self, fill_value=None):
             fill_value = self.fill_value
         else:
-            fill_value = _check_fill_value(fill_value, self.dtype)
+            if ((np.isscalar(fill_value) == True) or (self.shape == fill_value.shape)):
```
Its pretty tricky to do such a check. It probably makes sense to move this check into the `_check_fill_value` function. There are some tricky things about object arrrays (arrays with `dtype=object`). If we do not want to worry about it for now too much (it may be a different issue), what I think will work is: Adding an `fill_value.ndim == 0` check at the end of `_check_fill_value`. You could refactor the last `np.array` call, since it seems unnecessary except in the default fill value branch.

kritisingh1 replied:
If we keep that as a final check, a number of tests fail, which are passing non-scalar values as the fill_value. That's why, even in the above if statement, I let it be true even when the fill_value were non-scalar. as long as their shape was the same as that of the input array. So from here, either we have to modify a lot of tests or develop some other workaround. What do you suggest?

seberg replied:
oh OK, I missed that shape check. `fill_value` should be an array here though, so checking `.ndim == 0 or shape == fill_value.shape` may be an option? Depending on where else `_check_fill_value` is called, moving it inside and with a new shape argument may or may not make sense.

kritisingh1 commented on numpy/ma/core.py:
```diff
@@ -602,8 +602,10 @@ def filled(a, fill_value=None):
+    fill_value : array_like, optional. Can be
+        scalar or non-scalar. If non-scalar, the
+        resulting filled_array should be broadcastable
+        over input array. Default is None.
```
Should I write it like ``filled array`` @seberg ?

seberg replied:
I think just "filled array" without underscore or backticks since I do not think it is a name in code. (all tiny stuff though).

kritisingh1 committed:
minor fixes

kritisingh1 commented:
@seberg when I added a check like
> if((np.isscalar(fill_value) == True) or (ndtype.shape == fill_value.shape)):
>       return np.array(fill_value)
>  else:
>        err_msg = "Cannot find valid fill_value for dtype %s"
>        raise TypeError(err_msg % ndtype)
at the end of the `_check_fill_value`, around 18 tests are failing, logs of which are present [here](https://pastebin.com/FmHr86dR) Am I missing something?

seberg committed:
MAINT: Try different approach to deprecate non-scalar fill

Actually, we deprecate non-scalar or matching fills as it is there,
should check what we want to do.

mattip commented on numpy/ma/core.py:
```diff
@@ -3709,6 +3724,11 @@ def filled(self, fill_value=None):
+                warnings.warn(
+                    "Non-scalar arrays for fill values are deprecated. Use "
+                    "arrays with scalar values instead",
+                    DeprecationWarning, stacklevel=2)
```
Test needed. It should go into `numpy/ma/tests/test_deprecations.py` as a new class `TestFill` and simply `assert_warns(DeprecationWarning, ...)`

seberg replied:
I added a new commit moving the deprecation (it was on the wrong spot). So now setting the `masked_array.fill_value = something` should give a deprecation warning, but the filled function is unchanged. So that is what we need to test.

There is still a problem that it has to be added at least in line 2904 in the `__new__` method. Can you make that line spell `self.fill_value = fill_value` and write a test that it works.

The indexing code seems like it might be slightly wrong, but I think it is unrelated (the warning says it makes it 0D which is what we want). (so please ignore this last part)

seberg committed:
DEP: move deprecation from filled to fill_value

mattip commented on numpy/ma/core.py:
```diff
@@ -3648,6 +3659,13 @@ def fill_value(self):
+                "arrays with scalar values instead. The filled function "
+                "still supports any array as `fill_value`.",
+                DeprecationWarning, stacklevel=2)
+
```
Needs a date and version comment so we know when we can turn the deprecation into an error

mattip commented:
Release note needed for the new deprecation

rgommers commented:
This looks almost ready, just need to add a 1.18.0 version number to the deprecation, and a release note.

mattip committed:
DEP: add changelog note and comment in code

mattip commented:
Thanks @kritisingh1
